lib/spack/spack/solver/concretize.lp

```diff
@@ -366,6 +366,11 @@ attr(Name, node(X, A1), A2)         :- impose(ID, PackageNode), imposed_constrai
 attr(Name, node(X, A1), A2, A3)     :- impose(ID, PackageNode), imposed_constraint(ID, Name, A1, A2, A3),     imposed_nodes(ID, PackageNode, node(X, A1)), not multiple_nodes_attribute(Name).
 attr(Name, node(X, A1), A2, A3, A4) :- impose(ID, PackageNode), imposed_constraint(ID, Name, A1, A2, A3, A4), imposed_nodes(ID, PackageNode, node(X, A1)).
 
+attr(DepAttrName, DepNode, Value)
+   :- depends_on(node(X, Package), DepNode),
+      attr(PackageAttrName, node(X, Package), Value),
+      attr("sync", DepNode, DepAttrName, attr(PackageAttrName, Package)).
```

Comment:
I don't see how this realizes the `when` part of `depends_on("geant4-data@{my.version}", when="@10:")`

Reply:
when is already implemented -- the `sync` `attr` being derived is conditional on whatever condition id implements the `when`.

Reply:
given:
```python
    depends_on("version-lock-dep@my.version", when="@2:")
```

You get:
```prolog
> spack -m solve --show asp version-lock | grep -A 23 'condition(0)'
pkg_fact("version-lock",condition(0)).
condition_reason(0,"version-lock depends on version-lock-dep@my.version when @2:").
pkg_fact("version-lock",condition_trigger(0,1)).
pkg_fact("version-lock",condition_effect(0,2)).


%----------------------------------------------------------------------------
% Trigger conditions
%----------------------------------------------------------------------------
pkg_fact("version-lock",trigger_id(1)).
pkg_fact("version-lock",trigger_msg("version-lock@2:")).
condition_requirement(1,"node","version-lock").
condition_requirement(1,"node_version_satisfies","version-lock","2:").
condition_requirement(1,"track_dependencies","version-lock").


%----------------------------------------------------------------------------
% Imposed requirements
%----------------------------------------------------------------------------
pkg_fact("version-lock-dep",effect_id(2)).
pkg_fact("version-lock-dep",effect_msg("version-lock-dep@my.version")).
imposed_constraint(2,"sync","version-lock-dep","node_version_satisfies",attr("version","version-lock")).
imposed_constraint(2,"dependency_holds","version-lock","version-lock-dep","build").
imposed_constraint(2,"dependency_holds","version-lock","version-lock-dep","link").
```